src/main/scala/io/qbeast/spark/delta/writer/RollupWriteStrategy.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2021 Qbeast Analytics, S.L.
+ */
+package io.qbeast.spark.delta.writer
+
+import io.qbeast.core.model.IndexFile
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions.col
+import io.qbeast.IISeq
+import io.qbeast.spark.index.QbeastColumns
+import org.apache.spark.sql.catalyst.InternalRow
+import scala.collection.mutable
+import io.qbeast.core.model.CubeId
+import io.qbeast.core.model.TableChanges
+import io.qbeast.core.model.Weight
+
+/**
+ * Implementation of WriteStrategy that groups the records to write by "rolling"
+ * them up along the index tree.
+ *
+ * @param qbeastColumns the Qbeast-specisific columns
+ * @param tableChanges the table changes
+ */
+private[writer] class RollupWriteStrategy(
+    qbeastColumns: QbeastColumns,
+    tableChanges: TableChanges)
+    extends WriteStrategy
+    with Serializable {
+
+  override def write(
+      data: DataFrame,
+      writerFactory: IndexFileWriterFactory): IISeq[(IndexFile, TaskStats)] = {
+    data
+      .repartition(col(QbeastColumns.cubeColumnName))
+      .queryExecution
+      .executedPlan
+      .execute()
+      .mapPartitions(writeRows(writerFactory, targetCubeIds))
+      .collect()
+      .toIndexedSeq
+  }
+
+  private def writeRows(
+      writerFactory: IndexFileWriterFactory,
+      targetCubeIds: Map[CubeId, CubeId])(
+      rows: Iterator[InternalRow]): Iterator[(IndexFile, TaskStats)] = {
+    val writers: mutable.Map[CubeId, IndexFileWriter] = mutable.Map.empty
+    rows.foreach { row =>
+      val cubeId = getCubeId(row)
+      val targetCubeId = targetCubeIds.getOrElse(cubeId, cubeId)
+      val writer = writers.getOrElseUpdate(targetCubeId, writerFactory.newWriter())
+      writer.write(row)
+    }
+    writers.values.iterator.map(_.close())
+  }
+
+  private def getCubeId(row: InternalRow): CubeId = {
+    val bytes = row.getBinary(qbeastColumns.cubeColumnIndex)
+    tableChanges.updatedRevision.createCubeId(bytes)
+  }
+
+  private def targetCubeIds: Map[CubeId, CubeId] = {
```

Comment:
Assuming a small append (with a count smaller than the `desiredCubeSize`) on a large existing table. The existing weight for the root is expected to be small (say 0.1), and the appending estimated weight for the root, which is the only cube, will be >= 1.0.

In this case, the weight merge will create a small weight for the root, let's say 0.08, and will push all records with weight > 0.08, which is almost all appending data, to the rest of the existing index (see `CubeNormalizedWeights.mergeNormalizedWeights`). 

This leads to the creation of many small files, probably as many as the number of existing cubes, and these cubes will not be present in `tableChanges.cubeDomains`.

Reply:
@Jiaweihu08 thank you for the comment, could you please help me to understand which particular line of code is incorrect.